File: Kotlin/729_My_Calendar_I.kt
class MyCalendar() {
    private val events = TreeMap<Int, Int>()

    fun book(start: Int, end: Int): Boolean {
        val event = events.lowerEntry(end)

        if(event == null || event.value <= start) {
            events[start] = end
            return true
        }

        return false
    }

}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * var obj = MyCalendar()
 * var param_1 = obj.book(start,end)
 */

File: Cpp/729_My_Calendar_I.cpp
class MyCalendar {
public:
    MyCalendar() {
        
    }
    
    bool book(int start, int end) {
        ++events[start];
        --events[end]; 
        int sum = 0;

        for(auto& event : events) {
            sum += event.second;

            if(sum > 1) {
                --events[start];
                ++events[end];
                
                return false;
            }
        }
        
        return true;
    }
    
    map<int, int> events;
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */